File: src/modules/milestones.js
// ── milestones.js ────────────────────────────────────────────
// Projects current rate of progress onto goal timelines.
// All calculations in kg. Display via units.js.
// ─────────────────────────────────────────────────────────────

import { TARGETS, BASELINE } from './program.js'

/**
 * Given exercise history (array of { date, est1rm }),
 * project when the target 1RM will be reached.
 *
 * Uses a rolling linear regression over the last 8 data points
 * to smooth out noise from single bad sessions.
 *
 * @returns {object} { etaDate, etaLabel, ratePerWeek, weeksRemaining, pctComplete }
 */
export function projectMilestone (history, targetKg) {
  if (!history || history.length < 2) {
    return {
      etaDate:      null,
      etaLabel:     'Log more sessions',
      ratePerWeek:  0,
      weeksRemaining: null,
      pctComplete:  0,
    }
  }

  const recent = history.slice(-10)  // last 10 data points
  const n      = recent.length
  const first  = recent[0]
  const last   = recent[n - 1]

  const startKg    = first.est1rm
  const currentKg  = last.est1rm
  const startDate  = new Date(first.date)
  const lastDate   = new Date(last.date)
  const daysDiff   = Math.max(1, (lastDate - startDate) / (1000 * 60 * 60 * 24))
  const ratePerDay = (currentKg - startKg) / daysDiff
  const ratePerWeek = ratePerDay * 7

  const baselineKg  = BASELINE[exerciseFromHistory(history)] || startKg
  const totalGap    = targetKg - baselineKg
  const remaining   = targetKg - currentKg
  const pctComplete = totalGap > 0
    ? Math.min(100, Math.round(((currentKg - baselineKg) / totalGap) * 100))
    : 100

  if (ratePerDay <= 0) {
    return {
      etaDate:      null,
      etaLabel:     'Maintain consistency to project',
      ratePerWeek:  Math.round(ratePerWeek * 10) / 10,
      weeksRemaining: null,
      pctComplete,
    }
  }

  const daysRemaining   = remaining / ratePerDay
  const etaDate         = new Date(Date.now() + daysRemaining * 24 * 60 * 60 * 1000)
  const weeksRemaining  = Math.ceil(daysRemaining / 7)

  return {
    etaDate,
    etaLabel:      formatETA(etaDate, weeksRemaining),
    ratePerWeek:   Math.round(ratePerWeek * 10) / 10,
    weeksRemaining,
    pctComplete,
  }
}

function formatETA (date, weeks) {
  if (weeks <= 0)  return 'Target reached!'
  if (weeks <= 4)  return `~${weeks} week${weeks > 1 ? 's' : ''}`
  if (weeks <= 52) return date.toLocaleDateString('en-US', { month: 'long', year: 'numeric' })
  return 'Over 1 year at current rate'
}

// Attempt to infer exercise from history object if tagged
function exerciseFromHistory (history) {
  return history[0]?._exercise || null
}

/**
 * Generate projected weekly 1RM data points for chart overlay.
 * Returns array of { date (ISO string), kg } for the next N weeks.
 */
export function projectFuture (history, targetKg, weeksAhead = 26) {
  if (!history || history.length < 2) return []

  const recent      = history.slice(-6)
  const n           = recent.length
  const last        = recent[n - 1]
  const first       = recent[0]
  const daysDiff    = Math.max(1, (new Date(last.date) - new Date(first.date)) / (1000 * 60 * 60 * 24))
  const ratePerDay  = (last.est1rm - first.est1rm) / daysDiff

  if (ratePerDay <= 0) return []

  const points = []
  for (let w = 1; w <= weeksAhead; w++) {
    const projDate = new Date(new Date(last.date).getTime() + w * 7 * 24 * 60 * 60 * 1000)
    const projKg   = Math.round((last.est1rm + ratePerDay * w * 7) * 10) / 10
    if (projKg >= targetKg) {
      points.push({ date: projDate.toISOString().split('T')[0], kg: targetKg, isTarget: true })
      break
    }
    points.push({ date: projDate.toISOString().split('T')[0], kg: projKg })
  }
  return points
}

/**
 * Build mesocycle summary stats from session history.
 * Returns array of { meso, avgRPE, totalSets, prsHit, completionRate }
 */
export function mesoSummary (recentSessions) {
  const byMeso = {}
  for (const s of recentSessions) {
    const m = s.mesocycle || 1
    if (!byMeso[m]) byMeso[m] = { sessions: [], sets: [], prs: 0 }
    byMeso[m].sessions.push(s)
    for (const set of (s.sets || [])) {
      if (set.completed) {
        byMeso[m].sets.push(set)
        if (set.is_pr) byMeso[m].prs++
      }
    }
  }

  return Object.entries(byMeso).map(([meso, data]) => {
    const rpeSets  = data.sets.filter(s => s.rpe)
    const avgRPE   = rpeSets.length
      ? Math.round((rpeSets.reduce((a, s) => a + s.rpe, 0) / rpeSets.length) * 10) / 10
      : null
    const completed = data.sessions.filter(s => s.completed).length
    return {
      meso:            parseInt(meso),
      avgRPE,
      totalSets:       data.sets.length,
      prsHit:          data.prs,
      completionRate:  data.sessions.length
        ? Math.round((completed / data.sessions.length) * 100)
        : 0,
    }
  }).sort((a, b) => a.meso - b.meso)
}

File: streamlit_app/modules/milestones.py
from datetime import datetime, timedelta
from .program import BASELINE, TARGETS


def project_milestone(history: list, target_kg: float, exercise: str = None, user_baselines: dict = None) -> dict:
    if not history or len(history) < 2:
        return {
            "eta_date": None,
            "eta_label": "Log more sessions",
            "rate_per_week": 0,
            "weeks_remaining": None,
            "pct_complete": 0,
        }

    recent = history[-10:]
    first, last = recent[0], recent[-1]

    start_kg = first["est1rm"]
    current_kg = last["est1rm"]
    start_date = datetime.strptime(first["date"], "%Y-%m-%d")
    last_date = datetime.strptime(last["date"], "%Y-%m-%d")
    days_diff = max(1, (last_date - start_date).days)
    rate_per_day = (current_kg - start_kg) / days_diff
    rate_per_week = rate_per_day * 7

    effective_baselines = user_baselines or BASELINE
    baseline_kg = effective_baselines.get(exercise, start_kg) if exercise else start_kg
    total_gap = target_kg - baseline_kg
    pct_complete = (
        max(0, min(100, round(((current_kg - baseline_kg) / total_gap) * 100)))
        if total_gap > 0
        else 100
    )

    if rate_per_day <= 0:
        return {
            "eta_date": None,
            "eta_label": "Maintain consistency to project",
            "rate_per_week": round(rate_per_week * 10) / 10,
            "weeks_remaining": None,
            "pct_complete": pct_complete,
        }

    remaining = target_kg - current_kg
    days_remaining = remaining / rate_per_day
    eta_date = datetime.now() + timedelta(days=days_remaining)
    weeks_remaining = -(-int(days_remaining) // 7)  # ceiling

    return {
        "eta_date": eta_date,
        "eta_label": _format_eta(eta_date, weeks_remaining),
        "rate_per_week": round(rate_per_week * 10) / 10,
        "weeks_remaining": weeks_remaining,
        "pct_complete": pct_complete,
    }


def _format_eta(date: datetime, weeks: int) -> str:
    if weeks <= 0:
        return "Target reached!"
    if weeks <= 4:
        return f"~{weeks} week{'s' if weeks > 1 else ''}"
    if weeks <= 52:
        return date.strftime("%B %Y")
    return "Over 1 year at current rate"


def project_future(history: list, target_kg: float, weeks_ahead: int = 26) -> list:
    if not history or len(history) < 2:
        return []

    recent = history[-6:]
    first, last = recent[0], recent[-1]
    days_diff = max(
        1,
        (datetime.strptime(last["date"], "%Y-%m-%d") - datetime.strptime(first["date"], "%Y-%m-%d")).days,
    )
    rate_per_day = (last["est1rm"] - first["est1rm"]) / days_diff

    if rate_per_day <= 0:
        return []

    points = []
    last_date = datetime.strptime(last["date"], "%Y-%m-%d")
    for w in range(1, weeks_ahead + 1):
        proj_date = last_date + timedelta(weeks=w)
        proj_kg = round((last["est1rm"] + rate_per_day * w * 7) * 10) / 10
        if proj_kg >= target_kg:
            points.append({"date": proj_date.strftime("%Y-%m-%d"), "kg": target_kg, "is_target": True})
            break
        points.append({"date": proj_date.strftime("%Y-%m-%d"), "kg": proj_kg})

    return points


def meso_summary(recent_sessions: list) -> list:
    by_meso: dict = {}
    for s in recent_sessions:
        m = s.get("mesocycle", 1)
        by_meso.setdefault(m, {"sessions": [], "sets": [], "prs": 0})
        by_meso[m]["sessions"].append(s)
        for set_data in s.get("sets") or []:
            if set_data.get("completed"):
                by_meso[m]["sets"].append(set_data)
                if set_data.get("is_pr"):
                    by_meso[m]["prs"] += 1

    result = []
    for meso, data in by_meso.items():
        rpe_sets = [s for s in data["sets"] if s.get("rpe")]
        avg_rpe = (
            round(sum(s["rpe"] for s in rpe_sets) / len(rpe_sets) * 10) / 10
            if rpe_sets
            else None
        )
        completed = sum(1 for s in data["sessions"] if s.get("completed"))
        result.append({
            "meso": int(meso),
            "avg_rpe": avg_rpe,
            "total_sets": len(data["sets"]),
            "prs_hit": data["prs"],
            "completion_rate": (
                round(completed / len(data["sessions"]) * 100) if data["sessions"] else 0
            ),
        })

    return sorted(result, key=lambda x: x["meso"])
